features/202-session-management/api.md

```diff
@@ -0,0 +1,66 @@
+# Session Management APIs
+
+## SDK API
+```typescript
+interface UserAgent {
+    raw: string;
```

Comment:
only web have a "raw user agent"?

Reply:
I think exposing native SDK raw user agent string is acceptable?

Reply:
Does iOS / Android SDK's user agent string is meaningful at all? I guess not?

Reply:
It is meaningful, considering the parsed information is derived from the raw user agent string. Indeed it is unlikely that developer would use this directly, but exposing this should have no problem.

Reply:
Oh I guess a difference between us is I don't think we should use the user-agent string for all information. User agent string is just ONE source of information about the sessions' devices, a structured information could be passed from client SDK if it have additional information....

Reply:
I had discussed with the team on how much information should be included in session:
- To comply with GDPR, we should collect the minimum amount of information by default: preferably just the user agent & IP, since it can be justified for security purpose.
- I consider device name/ID is a personal identifier (e.g. John's iPhone), therefore we should not collect this by default.
- I'm not sure what you mean by 'additional information', but personally I think this is in grey area of user tracking and analytics.

Reply:
* Please look at the other conversation thread on what I meant when I say I don't think we should overload the user agent string. 
* IP address is personal identifier too, why IP address but not device name/ID? (However I do agree we can have a feature switch for device name/ID, but I guess it is definitely helpful for users to identify strange session quickly)

Reply:
As mentioned, user agent & IP can be justified for security purpose, and most server is already logging this information. However, device name contains potentially sensitive information (e.g. 'John's iPhone' would expose real name of user), and device ID is mostly useless for security purpose (cannot co-relate sessions on same device practically). It is much harder to justify the use of device name/ID.

Reply:
The justification is it is definitely helpful for users to identify strange session quickly

Reply:
> As mentioned the justification is it is definitely helpful for users to identify strange session quickly

If it is used to identify potential account intrusion automatically, or displayed in an end-user accessible UI, then likely it can be counted as 'legitimate purpose' under GDPR. However, if we do not have such automatic mechanism, and the UI is not implemented by developers, I think it would be a violation. (correct me if I'm wrong)

Reply:
Strictly speaking for GDPR, it doesn't matter if it is "legitimate" enough or not in these kinds of cases, what matter is the app need to explain and let user opt-in/opt-out if they want to do it or not.

But how granular the opt-in or out need to have yet to be defined, need more court case.

So technically to support GDPR, we need a feature switch for developers to turn this feature on/off for each user. Which we can create an issue and handle later if we want to stirctly follows and respect GDPR.

Reply:
If we handle it later, would it means the first official release would not be GDPR compliant?

I'm not a GDPR specialist, but I read that there are 6 basis which processing personal data is allowed. Opt-in is the 'consent' basis, and 'legitimate interest' is another basis, which security features would fall under, so consent is not required. Personally I find it hard to justify using device name/ID as security feature.

Reply:
@kiootic legally speaking a lot of things is a grey area for GDPR now. E.g. GDPR requires "granular" opt-in/out, how granular is an issue. Google thinks it is okay to make it like a few major areas to opt-in, Apple think it is okay to show one opt-in screen, and show an icon whenever privacy data is being collected... a lot of privacy-specialist think you need a long list for people to opt-in/out, some people try to file a suit to Google....

We will only know after the court makes the ruling.

I think for us we should use our best effort to enable support of GDPR compliant software, so for this feature, I would support if we add an feature issue that let people opt-in/out of session control. (probably too difficult to do it now and not common enough)

Similarly "Legitimate" is even more undefined, but generally speaking, it is legitimate interest if it is used for the benefit for that company. In this case, it is not relevant -- we are not even processing the data for the purposes of any benefits -- it is for a vital interest of the user -- to help them identify a potentially problematic session.

Reply:
> We will only know after the court makes the ruling.

However, it would be irresponsible to provide a feature enabled by default, and say it's developer's responsibility to check the ruling and see if need to ask for consent.

> I think for us we should use our best effort to enable support of GDPR compliant software, so for this feature, I would support if we add an feature issue that let people opt-in/out of session control. (probably too difficult to do it now and not common enough)

Yes, supporting per-user opt-in/out would not fit our current schedule. Therefore, we should  only enable features that we can justify the usage of personal data by default (i.e. IP and user agent as previously mentioned).

> Similarly "Legitimate" is even more undefined, but generally speaking, it is legitimate interest if it is used for the benefit for that company. In this case, it is not relevant -- we are not even processing the data for the purposes of any benefits

It a legitimate interest of developer to use user's personal data [for IT security](https://gdpr.eu/Recital-49-Network-and-information-security-as-overriding-legitimate-interest/), as permitted by GDPR.

> it is for a vital interest of the user -- to help them identify a potentially problematic session.

In context of GDPR, vital interest is [essential to life of the user](https://gdpr.eu/Recital-46-Vital-interests-of-the-data-subject/), which obviously is not applicable in our situation.

I'll suggest reading some summary on GDPR to understand more about it (personally used resources at [ICO](https://ico.org.uk/for-organisations/guide-to-data-protection/guide-to-the-general-data-protection-regulation-gdpr/lawful-basis-for-processing/)). I found it quite clear to understand.

Reply:
> However, it would be irresponsible to provide a feature enabled by default, and say it's developer's responsibility to check the ruling and see if need to ask for consent.

Common law just doesn't work that way. Until there is a court ruling even lawyer can't tell you what exactly the legal requirements is -- especially in tech-related law where legislator always have no clue lol.

I agree we can provide a best practices approach, but than it also relates with cost. For example we both agree support per-user opt-in will not fit our schedule.

I'm losing you here, what are we discussing?

Reply:
> I'm losing you here, what are we discussing?

We are discussing whether including device name/ID by default would be compliant to GDPR. As we discussed offline earlier, we could make including device name opt-in, since it would be beneficial to layman user for identifying devices. However, device ID is not need for now, so it is not included in session info.

Reply:
> In context of GDPR, vital interest is [essential to life of the user](https://gdpr.eu/Recital-46-Vital-interests-of-the-data-subject/), which obviously is not applicable in our situation.

BTW, I didn't use vital interest in legal terms here. So it is beyond the point of our discussion..

Reply:
In conclusion for offline discussion with @kiootic , unless there are more logical arguments:

1. we believe device name is a good information sources to help layman users to identify compromised sessions
1. device name is not accurate and can only be gather as a best effort
1. device name feature could be turned off or disable by default and turned on.